File: src/main/scala/xieyuheng/cicada/cicada.scala
package xieyuheng.cicada

import collection.immutable.ListMap

import xieyuheng.util.Interpreter
import xieyuheng.partech.Parser
import xieyuheng.partech.Earley
import xieyuheng.partech.ErrorDuringParsing

object cicada extends Interpreter {

  val name = "cicada"
  val version = "0.0.1"

  val config_declaration = ListMap(
    "--verbose" -> 0,
    "--nocolor" -> 0,
  )

  def run_code(code: String): Unit = {
    val earley = Earley()
    val parser = Parser(grammar.lexer, grammar.top_list, earley)
    try {
      val tree = parser.parse(code)
      val top_list = grammar.top_list_matcher(tree)
      api.run(top_list, this.config)
    } catch {
      case error: ErrorDuringParsing =>
        println(s"parsing error, at span: (${error.span.lo}, ${error.span.hi})")
        println(s"${error.message}")
        System.exit(1)
    }
  }

}

File: src/cicada.ts
import * as pt from "@forchange/partech"
import * as Top from "./top"
import * as CLI from "./cli"
import * as API from "./api"
import * as grammar from "./grammar"

const pkg: any = require("../package.json")

function run_code(code: string, config: { [key: string]: any }): void {
  const lexer = pt.lexers.common
  const parser = pt.parsers.earley

  try {
    let tokens = lexer.lex(code)
    let tree = parser.parse(tokens, grammar.top_list())
    let top_list: Array<Top.Top> = grammar.top_list_matcher(tree)
    API.run(top_list, config)
  } catch (error) {
    if (error instanceof pt.Err.Parsing) {
      console.log(`parsing error, at ${pt.Span.repr(error.span)}`)
      console.log(pt.Span.repr_in_context(error.span, code))
      console.log(`${error.message}`)
      process.exit(1)
    } else {
      throw error
    }
  }
}

export const cli = new CLI.CommandLineInterface(pkg.name, pkg.version, run_code)
